From 437ffc6e1099219458b1daf8458417539a331d59 Mon Sep 17 00:00:00 2001
From: Liu Bolun <bolunx.liu@intel.com>
Date: Tue, 11 Sep 2012 11:29:48 +0800
Subject: Adding free heap when start encode fail.

BZ: 56455

we should hanlde error occur in initialization that calling
MediaSource::stop()when mState equals to LOADED_TO_IDLE,
it means OMXComponet is initializing only in initialization stage.
In addition, we not only stop mediasource, but also free buffer for
OMXCodec ports avoid memory leak.Modifying code style.

Category: aosp improvement
Domain: Video.Media-Framework
Origin: Internal
Upstream-Candidate: Yes

Change-Id: I1fe805721cc9ed65d80432faf034328be0725843
Orig-Change-Id: Ia7869bfd8f01da82d07502fab7956152350fd650
Signed-off-by: Liu Bolun <bolunx.liu@intel.com>
Signed-off-by: Dan Liang <dan.liang@intel.com>
---
 media/libstagefright/OMXCodec.cpp |   10 ++++++++++
 1 files changed, 10 insertions(+), 0 deletions(-)

diff --git a/media/libstagefright/OMXCodec.cpp b/media/libstagefright/OMXCodec.cpp
index 76c394c..6cd16f5 100644
--- a/media/libstagefright/OMXCodec.cpp
+++ b/media/libstagefright/OMXCodec.cpp
@@ -2750,6 +2750,16 @@ void OMXCodec::onEvent(OMX_EVENTTYPE event, OMX_U32 data1, OMX_U32 data2) {
         {
             CODEC_LOGE("ERROR(0x%08lx, %ld)", data1, data2);
 
+            // we should hanlde error occur in initialization that calling MediaSource::stop()
+            // when mState equals to LOADED_TO_IDLE, it means OMXComponet is initializing
+            // only in initialization stage.
+            if(mState == LOADED_TO_IDLE)
+            {
+               freeBuffersOnPort(kPortIndexInput);
+               freeBuffersOnPort(kPortIndexOutput);
+               mSource->stop();
+            }
+
             setState(ERROR);
             break;
         }
-- 
1.7.4.1

